Doc/library/typing.rst

```diff
@@ -585,6 +587,31 @@ These can be used as types in annotations and do not support ``[]``.
    .. versionadded:: 3.5.4
    .. versionadded:: 3.6.2
 
+.. data:: Self
+
+   Special type to represent the current enclosed class.
+   For example::
+
+      from typing import Self
+
+      class Foo:
+         def returns_self(self) -> Self:
+            ...
+            return self
+
+
+   This annotation is semantically equivalent to using a :class:`TypeVar` with ``bound=Foo`` as
+   both the return annotation and the annotation for the ``self`` parameter.
+
+   Common use cases include:
+
+      - :class:`classmethod`\s that are used as alternative constructors
```

Comment:
Isn't that debatable? E.g. `float.fromhex`, `dict.fromkeys` don't promise this -- not all alternative constructors are this smart. So maybe it needs clarification? Or is the mention of "class method" sufficient?